src/Open3D/GUI/Slider.cpp

```diff
@@ -0,0 +1,132 @@
+// ----------------------------------------------------------------------------
+// -                        Open3D: www.open3d.org                            -
+// ----------------------------------------------------------------------------
+// The MIT License (MIT)
+//
+// Copyright (c) 2018 www.open3d.org
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+// IN THE SOFTWARE.
+// ----------------------------------------------------------------------------
+
+#include "Slider.h"
+
+#include "Theme.h"
+
+#include <imgui.h>
+
+#include <algorithm>
+#include <cmath>
+#include <sstream>
+
+namespace open3d {
+namespace gui {
+
+namespace {
+static int gNextSliderId = 1;
+}
+
+struct Slider::Impl {
+    Slider::Type type;
+    std::string id;
+    // A double has 53-bits of integer value, which should be enough for
+    // anything we want a slider for. A slider isn't really useful for
+    // a range of 2^53 anyway.
+    double value = 0.0;
+    double minValue = -1e35;
```

Comment:
why `NumberEdit` (from `-2e9` to `2e9`) has different default range, will the larger range here be problematic for `ImGui::SliderInt`?

Reply:
The ranges will work okay, and they can be set to any valid `double` value. I think they are different because a) I probably wrote the files a different times, and b) I think of NumberEdit as going to be numbers used for something concrete and but I think I was thinking of a Slider being more of a floating point selector, and 1e35 is likely to be more than anyone is going to need for anything remotely physical (I think the diameter of the Universe is ~1e29 millimeters); so effectively the slider has no default min/max. If you think it better, I can change it to something else.